torch/_functorch/aot_autograd.py

```diff
@@ -1204,6 +1205,16 @@ def inner(*flat_args):
             [x for x in input_info if x.mutates_data or x.mutates_metadata]
         )
 
+        # If a subclass does extra aliasing between graph outputs/inputs in a way
+        # that is not visible above the sublass, then we can end up having tangents
+        # that alias inputs.
```

Comment:
nit: Can you add more details to this note?

Ordinarily, when constructing the joint function that we want to trace in AOTAutograd, we're guaranteed that we the tangent tensors that we pass into the joint are distinct tensors from the primals. This is because when decide which forward outputs to create tangents for, we **only** create tangents for forward outputs that are not aliases of inputs (See Note [AOT Autograd: outputs aliasing inputs or intermediates!]).

However, when wrapper tensor subclasses enter the picture, it is possible to have an output of the forward that is a subclass that is not an input / alias of an input, **but** one of its inner tensors is an alias! NestedTensor is an example: Performing an out-of-place pointwise op on a NestedTensor will often construct a fresh NestedTensor that holds onto the input's offsets tensor directly.

Ordinarily, we can just pass in the traced (fake tensor) fw outs directly as our tangents when tracing the joint. But if we have tangent tensors that **are** the same as the (primal) forward inputs, this can cause problems during tracing: make_fx() will specialize on our duplicate inputs. If we passed in the same tensor for `primals_1` and `tangents_1` during tracing, `make_fx()` will happily sub out all usages of `tangents_1` with `primals_1` in the graph, which is not what we want.

To work around this, we view every forward output when creating out tangent tensors.

Reply:
Thanks for the detailed write up! I've included the details into the note and moved it to the top of the file with the other notes to avoid breaking the flow of the code since its a little longer now (let me know if you prefer it inline).